### DISCUSSION TITLE
Enable clang-tidy for clangd

### DISCUSSION THREAD

**Discussion by tdejager** [williamboman/nvim-lsp-installer#392] [Q&A] [answered]
Hey!

First of all, I absolutely love this plugin! Thanks for making this.

I have a question regarding the clangd lsp server. I want to enable clang-tidy support for this lsp. However, this is done through a command line flag as seen here: https://neovim.discourse.group/t/how-to-enable-the-header-insertion-of-clangd/1234 . As opposed to a json setting, I believe.

To do this I would need to override the `cmd` variable that is passed to the `server:setup{}` function. However, the actual path is abstracted away by this plugin. How could I get this path *or* add the flag to the server command somehow?

Thanks!

**Comment by williamboman** [comment, marked as answer]
Hello and thanks for your kind words! Recently I made a change where nvim-lsp-installer no longer (apart from a few exceptions) overrides the `cmd` governed by lspconfig. Instead it will extend the `PATH`, allowing you to configure the `cmd` accordingly. So you should be able to just

```lua
server:setup {
  cmd = {
     "clangd",
     "--background-index",
     "-j=12",
     "--query-driver=/usr/bin/**/clang-*,/bin/clang,/bin/clang++,/usr/bin/gcc,/usr/bin/g++",
     "--clang-tidy",
     "--clang-tidy-checks=*",
     "--all-scopes-completion",
     "--cross-file-rename",
     "--completion-style=detailed",
     "--header-insertion-decorators",
     "--header-insertion=iwyu",
     "--pch-storage=memory",
   }
}
```

**Reply by tdejager** [reply]
Thanks! Hmm, but that does not seem to work for me. With the newest version I get the error:

```
Spawning language server with cmd: `clangd` failed. The language server is either not installed, missing from PATH, or
 not executable.
```

**Reply by williamboman** [reply]
Have you updated the plugin version recently? That change was added just the other day

**Reply by williamboman** [reply]
Oh and you also have to reinstall clangd as the installation directory structure was changed as well

**Reply by tdejager** [reply]
Yeah, re-installing clangd dit the trick. I'm stil not getting the clang-tidy diagnostics, but a quick `ps aux | grep clangd` did show the correct flags being passed. So I suppose it is a different problem. Let's consider this solved :)

**Reply by kelvin-van-vuuren** [reply]
@tdejager did you manage to get this working? I'm in the same situation.` ps aux | grep clangd` is showing correct flags but no diagnostics showing up in neovim. However I can use a .clangd config file at project root and run clangd without any of the above flags and diagnostics will show up. Though if possible I'd much rather have this enabled via nvim lua config

**Reply by ravenxrz** [reply]
@tdejager @kelvin-van-vuuren Try clangd 12.0.0. diagnostics from clangd-tidy works well without .clang-tidy file.

**Comment by tdejager** [comment]
@kelvin-van-vuuren I saw that the `--clang-tidy-checks` flag was deprecated in the latest clangd release. However, I read that if you provide a `.clang-tidy` file that it should work. I hadn't had time to try it yet, was using Clion in the mean time (blasphemy I know :imp:). Not sure what the contents of the file should be, maybe you could try it and report the results here :)?

**Reply by c60cb859** [reply]
It works with a .clang-tidy file in you project dir
![image](https://user-images.githubusercontent.com/968766/158027509-47812503-f964-4ff1-b4ab-820016745a89.png)

This is my standard .clang-tidy file, it is stolen more or less from Jason Turners starter project https://github.com/cpp-best-practices/cpp_starter_project/blob/main/.clang-tidy
clang has greate documentation about how to add and remove checks with `+` or `-` the check name

```---
Checks: "*,
        -abseil-*,
        -altera-*,
        -android-*,
        -fuchsia-*,
        -google-*,
        -llvm*,
        -modernize-use-trailing-return-type,
        -zircon-*,
        -readability-else-after-return,
        -readability-static-accessed-through-instance,
        -readability-avoid-const-params-in-decls,
        -cppcoreguidelines-non-private-member-variables-in-classes,
        -misc-non-private-member-variables-in-classes,
"
WarningsAsErrors: ''
HeaderFilterRegex: ''
FormatStyle:     none
```
